### PR TITLE
Replace incorrect wasm compilation fallback logic. Fixes #42055 partly

### LINKED ISSUE #42055: WebAssembly compilation fallback logic is wrong. May need some further docs about configuring memory limits.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I developed my own PWA application using Blazor. The app works normally on my computer and phone. But on my second phone the application doesn't have time to load and an error appears.
I connected my phone to the pc and through the chrome devtools I went to see the log on the console and I found the following errors.
```
Streaming compilation failed. Falling back to ArrayBuffer instantiation.  RangeError: WebAssembly.instantiate(): Out of memory: wasm memory

blazor.webassembly.js:1 TypeError: Failed to execute 'arrayBuffer' on 'Response': body stream already read
(anonymous)	@	blazor.webassembly.js:1
(anonymous)	@	blazor.webassembly.js:1
await in (anonymous) (async)		
(anonymous)	@	blazor.webassembly.js:1
createWasm	@	dotnet.6.0.2.hy3rmx5lfa.js:1
(anonymous)	@	dotnet.6.0.2.hy3rmx5lfa.js:1

blazor.webassembly.js:1 Uncaught (in promise) TypeError: Failed to execute 'arrayBuffer' on 'Response': body stream already read
    at blazor.webassembly.js:1:36928
    at async blazor.webassembly.js:1:36900
    at async blazor.webassembly.js:1:36636
(anonymous)	@	blazor.webassembly.js:1
await in (anonymous) (async)		
(anonymous)	@	blazor.webassembly.js:1
createWasm	@	dotnet.6.0.2.hy3rmx5lfa.js:1
(anonymous)	@	dotnet.6.0.2.hy3rmx5lfa.js:1
```

The PWA is distributed as AzureStaticApps.
Some specifics about the phone having these problems:
- Phone model: Samsung Galaxy J5 (2016)
- Android version: 7.1.1
- Chrome version: 98.0.3758.87
- Edge version: 97.0.1702.78

I attach some photos of the console and network traffic

Schermata 2022-02-09 alle 11 09 33
Schermata 2022-02-09 alle 11 09 53
Schermata 2022-02-09 alle 11 10 04
Schermata 2022-02-09 alle 11 16 02

### Expected Behavior

The PWA should work just as well as it does on the PC and other phones

### Steps To Reproduce

Simply, create a PWA project and try it out on low-end phones.

### Exceptions (if any)

_No response_

### .NET Version

6.0.102

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
Previously we:

 * Attempted to compile using `instantiateStreaming`
 * ... and if that failed, attempted to compile using `instantiate`

The problem is that this kind of fallback has never worked, because the first call to `instantiateStreaming` consumes the response body, so the second step would be guaranteed to fail in all cases.

We could solve this using `response.clone()`, but there's no known use case for the fallback anyway. As long as we only attempt to use `instantiateStreaming` when the content-type `application/wasm` is found (as per [spec](https://webassembly.github.io/spec/web-api/#dom-webassembly-instantiatestreaming)), then it should work, and if not there's no reason to think that `instantiate` would be more successful.

So as of this PR, we don't do fallback - we just pick between the two compilation APIs based on the content-type. This should increase our compatibility versus what we did before.

Fixes one of the main concerns raised in https://github.com/dotnet/aspnetcore/issues/42055

### CODE REVIEW

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts [hunk 1]
```diff
@@ -652,24 +652,28 @@ async function loadICUData(icuDataResource: LoadingResource): Promise<void> {
...
+  if (hasWasmContentType && typeof WebAssembly.instantiateStreaming === 'function') {
+    // We can use streaming compilation. We know this shouldn't fail due to the content-type header being wrong,
+    // as we already just checked that. So if this fails for some other reason we'll treat it as fatal.
+    const streamingResult = await WebAssembly.instantiateStreaming(wasmResourceResponse, imports);
+    return streamingResult.instance;
+  } else {
+    if (!hasWasmContentType) {
+      // In most cases the developer should fix this. It's unusual enough that we don't mind logging a warning each time.
+      console.warn('WebAssembly resource does not have the expected content type "application/wasm", so falling back to slower ArrayBuffer instantiation.');
     }
```
Comment (TanayParikh) [inline src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:670]: > we only attempt to use instantiateStreaming when the content-type application/wasm is found (as per [spec](https://webassembly.github.io/spec/web-api/#dom-webassembly-instantiatestreaming)), then it should work, and if not there's no reason to think that instantiate would be more successful.

In that case, should we just throw an exception given we don't expect the subsequent array buffer instantiation to work if `!hasWasmContentType`?

Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:670]: Non-streaming compilation should work regardless of the content type. It’s only streaming compilation that has the content type prerequisite.

Comment (SteveSandersonMS) [reply src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:670]: I think the way I phrased the code comment was ambiguous and perhaps misleading. I’ll try to make the comment clearer :)

**Review by TanayParikh — APPROVED** [review]
Thanks for the clarification! LGTM

**Pull request merged**